### ISSUE TITLE
Treat JPMS module names a first-class library properties

### ISSUE THREAD

**Issue by sormuras** [sonatype-nexus-community/search-maven-org#6]
JPMS was introduced last year with Java 9. By now, many library authors claimed a well-known name for their libraries. Either by specifying an `Automatic-Module-Name` in the `MANIFEST.MF` of the jar file or by an explicit `module-info.class` entry.

When using a library via JPMS in a modular project, the end-user has to know that specific module name. As far as I know, there's no repository, search-engine or meta-crawler that exposes that information to the public.

Having an automatically updated index that makes repositories searchable for JPMS module names would be great to have.

### tl;dr
Supersede @jodastephen 's https://github.com/jodastephen/jpms-module-names -- which offers a semi-automatic updated table of module name to group:artifact relation overview.

**Comment by rgra** [comment]
I second that thought.

To be able to move forward to modular projects it is neccesary that developers:
1) Know if a maven artifact is Java 9 ready with
 a) Having an 'Automatic-Module-Name' which makes them consumable as Java 9 Module
 b) Have a module-info.class which makes them fully Java 9 compatible by allowing JLinker to work
 which could be shown as a "Java 9" flag in two different outcomes along the artifact info in the beta UI.
2) Be able to search for a dependency by artifact name or module name.
3) Get a feeling for how many projects are already compatible, by having a list showing which dependencies are ready with respect to 1a and 1b

I get asked a lot about those points when giving Java 9 releated talks.

mvn dependency:resolve 
already shows the module names for all of the dependencies when running it on Java 9 but it doesn't seem to make a distinction between automatic module names from Jar name or by 'Automatic-Module-Name' entry (still have to verify this though).

Because it is resource consuming for the community to mirror full maven central and compute the properties ourselves to get a list which is alway up to date, it would be very nice if sonatype could add some of those features to maven central.

See also the thread at:
https://twitter.com/rgransberger/status/987267392878403584

**Comment by sormuras** [comment]
Hi @rgra

I explained this idea in more detail at https://gitter.im/sonatype-nexus-community/search-maven-org -- want to join the chat there? I will also expand the intial description with information scattered around some blog posts and other issues. Seems like a good place to collect the bits.

While we're discussing "the automated and central-ized solution", do you have some more libraries that are JPMS-ready and not listed [here](https://github.com/jodastephen/jpms-module-names/tree/master/generator/table)? Please, tell me about them by opening an issue or a PR. :)

My vision is that one day you may use JPMS module names to specify 3rd-party dependencies. A future build tool feature could also examine your `module-info.java` files and resolve all `requires` statements. Given that you somehow tell the tool which version you want. Via a [`module-version.properties`](https://sormuras.github.io/blog/2017-12-13-module-info-properties.html) perhaps? I already have a working proof-of-concept [Gradle plugin](https://sormuras.github.io/blog/2018-01-19-mod2mav.html) that uses both, the hand-crafted `module-maven.properties` and a local `module-version.properties`, to achive the goal: _use module names only_

**Comment by gunnarmorling** [comment]
Hi,

> My vision is that one day you may use JPMS module names to specify 3rd-party dependencies

I don't think that's a generally viable approach because there's not necessarily a 1:1 relationship between a "logical" module (i.e. a defined set of packages and classes) and a JAR artifact providing that module. E.g. in case of JCP/JarkartaEE etc. specifications, there may be multiple artifacts providing one and the same API, e.g. Hibernate's and Apache's versions of the JPA or Bean Validation APIs. They all must share the same module name (in the BV 2.0 spec, we recommend one, in future versions it'd be a mandatory part of the spec), so to be interchangeable. Ideally, there'd indeed be just a single artifact for such API which is shared across implementations, but it's not something that's going to be guaranteed.

**Comment by sormuras** [comment]
>  Ideally, there'd indeed be just a single artifact for such API which is shared across implementations, but it's not something that's going to be guaranteed.

Then please extract the API into a module and share it under a unique name (and jar). :)

Do I read your example correctly, that when I write in my ideal world:
`requires org.hibernate.bv` I would get a jar that contains two "logical" modules, one being "javax.jpa"?
When I declare instead `requires org.apache.bean.validation` I'd get also two "logical" module as well, also on being "javax.jpa"?

**Comment by jodastephen** [comment]
My view is that search.maven.org and other search engines should allow search by module name. The results of a search by module name would be a list of artifacts that match (not just one). There may be multiple in the case of EE (outlined by Gunnar) or if a project has been patched and the patch published.

It is not wrong for maven central to have multiple artifacts with the same module name, but we do need a way to find them. This means that future versions of Maven will still need groupId:artifactId, as it is not up to Maven central to define what the best mapping is from a module name to an artifact.

**Comment by forax** [comment]
I think that Stephen and Christian are both right :)
At least, you should be able to search for a specific module by its Java name (JPMS) and you should see if a specific artifact as a module-name, a module-info.class or nothing.

but that's not enough, we start seeing artifact that are wrongly modularized, i.e. the module-info and the POM disagree on the dependencies (log4j2 had that bug 2 or 3 releases ago), so considering the POM as the source of truth, we need to verify that all the module required in the module-info are listed in the dependencies. This can be done by Maven or Nexus.

**Comment by khmarbaise** [comment]
@forax Hm..Maybe I misunderstand your concern related to POM relation to modules and dependencies. What about test dependencies in the pom which are usually not part of the JPMS modules (require?) ? Furthermore that might not work cause you can have multi-version jar's which are working for JDK 8, 9+...so thats a requirement which will not work...

**Comment by forax** [comment]
The idea is to get the required java modules from the project module-info and see if there is a corresponding module-info with the same name in one of the dependencies of the POM.
If there is no project module-info to check :)

A required module can be declared static which means it is optional at runtime, so it should be either a normal dependencies or an optional dependencies.

Maven doesn't support to have different module-infos one for the code and one for the test, so you can not validate anything for the test.

Bascially, help people that know how to write a POM file to write a correct module-info.java.

**Comment by karianna** [comment]
As an FYI - we're starting some research on this at AdoptOpenJDK - See the [kick off document](https://docs.google.com/document/d/1rT9tWLdH0zPbQaZ6-eAJ5NKU6NI78ATcsXSWhLYhkw8/edit?usp=sharing) for details.

**Comment by sormuras** [comment]
@jorlina, @hboutemy, Terry and me just met online and discussed next steps to get this feature into search.maven.org -- we'll continue in about 2 weeks.

In the meantime, @jodastephen's `jpms-module-names` project was superseded by https://github.com/sormuras/modules. That _database_ collected already over **4.000 unique module names** by parsing the published results produced by an instance of https://github.com/sandermak/modulescanner running at Sonatype's infrastructure. The [sormuras/modules](https://github.com/sormuras/modules) database is updated on a daily cadence to include the latest modular libraries published on Maven Central.

Looking forward to see Java module names, like `org.objectweb.asm`, pop up here...
![image](https://user-images.githubusercontent.com/2319838/58195562-83ea8900-7cc8-11e9-9cb6-249179c58d84.png)

**Comment by sormuras** [comment]
The new "versions overview result page", for example for https://search.maven.org/artifact/de.sormuras.bach/de.sormuras.bach, could also show the name of the Java module provided by a JAR artifact:

![image](https://user-images.githubusercontent.com/2319838/69916992-e257a780-1461-11ea-9cbc-5563566eb44e.png)

Just between "Version" and "Updated". ;-)

**Comment by sormuras** [comment]
FYI: Created a similar feature request at https://youtrack.jetbrains.com/issue/PKGS-451 for Jetbrains' new Package Search product.

Here's a similar query https://package-search.jetbrains.com/package?id=de.sormuras.bach:de.sormuras.bach with a screenshot (w/o Java module API information):

![image](https://user-images.githubusercontent.com/2319838/78580613-f0b03880-7832-11ea-9423-e135be96f1d8.png)

**Comment by sormuras** [comment]
_ping_

**Comment by sormuras** [comment]
_pong_

**Comment by DarthHater** [comment]
Just an FYI, @jorlina and I got this in progress

**Comment by DarthHater** [comment]
@sormuras how would you prefer search work? That's not immediately clear to me. Displaying the info once you've found something I am figuring out, but when you go to search, how do you want to search for something on a java module name?

**Comment by sormuras** [comment]
> how would you prefer search work?

Not sure, if this answers your question, but here are some example searches.

### Search for all artifacts with this exact Java module name

- `m:junit`
- `m:org.junit.jupiter`

### Search for all artifacts that coordinates (group, artifact, module) contain this string

- `junit`
- `org.junit.jupiter`

Basically, treat Java module names like the other coordinates as explained here:
![image](https://user-images.githubusercontent.com/2319838/121653884-7108a400-ca9d-11eb-82d1-f16f5ee00256.png)

**Comment by sormuras** [comment]
See also https://github.com/sonatype-nexus-community/search-maven-org/issues/6#issuecomment-384446598 and https://github.com/sonatype-nexus-community/search-maven-org/issues/6#issuecomment-385181119.

**Comment by sormuras** [comment]
> The Java module system was introduced last year with Java 9. [...]

Java 17 is coming!

**Comment by sormuras** [comment]
Is adding Java modules names to the read-only UI part of the "few other important milestones that will precede this work." mentioned here? 🤔 

https://blog.sonatype.com/maven-central-and-sigstore

**Comment by sormuras** [comment]
> > The Java module system was introduced last year with Java 9. [...]
> 
> Java 17 is coming!

Java 19 is coming... 😉

Watching now https://www.youtube.com/watch?v=m9zHea6w6gs

**Comment by sormuras** [comment]
> Java introduced modules years ago with Java 9.

Updated the initial description because Java 20 is around the corner.

**Comment by emaphis** [comment]
Now, Java 21 is here.

**Comment by sgammon** [comment]
Hello from Java 22

**Comment by sormuras** [comment]
Hello from Java 23 in 2025 🤓

**Comment by sgammon** [comment]
@sormuras I [started on an experiment for this](https://github.com/javamodules/javamodules.dev), but abandoned it due to lack of time and some challenges regarding database size. Maybe I could find some time to work on it if others are interested, though, and the domain `javamodules.dev` is available for use. I would also be happy to contribute that back to Sonatype for community use if this feature ever lands.

**Comment by sgammon** [comment]
@sormuras There are some [NPM tools](https://www.npmjs.com/search?q=%40javamodules) which make it easier now, and those are well tested, and it turns out it could be possible to parse module names directly from JAR downloads within pure JS. This would make the site hostable in a cheap Cloudflare Worker without need for a persistent backend. Anyway, food for thought.
